src/stirling/bpf_tools/task_struct_resolver.cc

```diff
@@ -137,8 +137,54 @@ StatusOr<TaskStructOffsets> ScanBufferForFields(const struct buf& buf,
 
   TaskStructOffsetsManager offsets_manager(&task_struct_offsets);
 
+#if AARCH64
+  // On aarch64, the cpu_context struct inside the thread_struct, can contain the same address as
+  // the group_leader value of the task_struct. Since the thread_struct is always at the end of the
+  // task_struct, we can just stop searching when we get to the thread_struct. To identify the
+  // offset of the thread_struct, we search for the address of thread local storage which is always
+  // stored `sizeof(cpu_context) + some padding` after the beginning of the thread struct.
+  uint64_t tls_addr;
+  // ARM magic to get the system register TPIDR_EL0, which stores the address of thread local
+  // storage.
+  asm volatile("mrs %0, TPIDR_EL0" : "=r"(tls_addr));
+  VLOG(1) << "TLS address = " << tls_addr;
+  // For now we iterate through the buffer twice. At some point we should refactor offsets_manager
+  // to eliminate the need to iterate through the buffer twice.
+  int stop_at_offset = -1;
   for (const auto& [idx, val] : Enumerate(buf.u64words)) {
     int current_offset = idx * sizeof(uint64_t);
+    if (val == tls_addr) {
+      VLOG(1) << absl::Substitute("[offset = $0] Found tp_value", current_offset);
+      // We leave some room for padding. The thread_struct almost always seems to have some padding
+      // between cpu_context and tp_value. In the case that there is no padding, it shouldn't affect
+      // us since the values we care about are never right before the thread_struct in the
+      // task_struct.
+      constexpr int padding = 8;
+      // aarch64's cpu_context struct in linux stores 13 registers. I checked this in v4.14 and v6.1
+      // and it hasn't changed.
+      constexpr int num_cpu_context_registers = 13;
+      constexpr int tp_value_offset_in_thread_struct =
+          padding + num_cpu_context_registers * sizeof(uint64_t);
+      stop_at_offset = current_offset - tp_value_offset_in_thread_struct;
+      VLOG(1) << "Ignoring thread_struct at offset = " << stop_at_offset;
+      break;
+    }
+  }
+  if (stop_at_offset == -1) {
+    return error::Internal(
+        "Failed to find the thread_struct offset within the task_struct. This is required for "
+        "resolving task struct offsets on aarch64");
+  }
+#endif
+
+  for (const auto& [idx, val] : Enumerate(buf.u64words)) {
+    int current_offset = idx * sizeof(uint64_t);
+
+#if AARCH64
```

Comment:
And I see waht you mean about iterating over the buffer twice. Perhaps we should inline the ARM logic right now?

Reply:
So I think the best way to do this would be to change the OffsetsManager to keep around a list of candidates for each offset, have a function to set an ignored region of offsets on the offsets manager, and then resolve the candidate lists to a single offset at the end. 

However, I kind of want to defer this change for now because its a pretty big change to the way the OffsetsManager works. Additionally, my current diff has the nice property that there's zero change to behaviour on x86_64, so there's no worry about regressions there.